feat: semantic conventions via Weaver

hendriknielaender opened the pull request:
This PR introduces comprehensive Zig language support for OpenTelemetry semantic conventions, utilizing [weaver](https://github.com/open-telemetry/weaver/tree/main) template registry. The implementation includes:

• Complete template system for generating Zig source files from semantic convention specifications
• All four semantic convention modules: `attributes, metrics, resources and traces`

hendriknielaender committed:
feat: add OpenTelemetry semantic conventions support

hendriknielaender committed:
chore: rename to PROJECT_DIR

hendriknielaender commented:
@inge4pres what do you think about this approach? Mostly transfered as well from the Rust OTel approach. I saw that you also merged another approach to generated these.

inge4pres commented:
> @inge4pres what do you think about this approach? Mostly transfered as well from the Rust OTel approach. I saw that you also merged another approach to generated these.

@hendriknielaender apologies for not seeing this earlier, just figured out now you did this PR beacuse I don't have yet filters on this repo 😅 

Let me have a look, my approach is completely different but it's limited to regitry and much more complex to reason about because it gnerates code with `print` statement, while you are suggesting to adoopt templating which seems much nicer.

I'll try it out and see the results and let you know 👍🏼

inge4pres commented:
@hendriknielaender I've had time to review the project structure, how it works and the motivations.
btw thanks so much for the contribution, much appreciated as always 🙏🏼 

Here's a summary of differences with the approach I developed in the past weeks, and pros and cons that I found:

### Code generation

The approach in this PR based on Weaver definitely reconnects better with the OTel ecosystem than a custom generator I wrote.
The YAML parser we currently have is also unable to parse all the flavors of comments, leaving some folders not generated.

Maintaining less code is better in this case.

### Architecture

With this PR we lose the namespacing, all the attributes/entities are in a single file for all different folders.
On top of that, the attributes are treated as simple `[] const u8` with their name, so we lose all the metadata such as stability level and examples that instead we get when attached to real types.

I see that all of these metadata are splatted in comments, and while this is addressing the DX in the IDE, it's not allowing for usage based on those metadata.

### Summary

The proposed usage of Weaver is a good thing for the evolution of the semconv development in Zig.
I think it would be nice to steer into the direction you propose, as long as:

* we can maintain a richer type structure that allows to get all the metadata of attributes/entities via struct fields
* we can keep separate namespaces for each "module"
* we can code in build.zig a step to render all (or parts of) the content from the official semconv repo

Questions:

1. is it possible to instruct the weaver tool to generate only a subset of the convention? this would allow namespacing
2. how difficult is it to create richer templates that can populate structs based on which input we are given?

inge4pres committed:
port types and restart attributes

Signed-off-by: inge4pres <fgualazzi@gmail.com>

inge4pres commented:
@hendriknielaender just so you know, I am trying to iterate on your PR to add the characteristics that i mentioned in my previous comment

inge4pres committed:
apply rich types to metrics and resource

Signed-off-by: inge4pres <fgualazzi@gmail.com>

inge4pres committed:
apply rich types to trace

Signed-off-by: inge4pres <fgualazzi@gmail.com>

inge4pres committed:
zig fmt and delete older folders

Signed-off-by: inge4pres <fgualazzi@gmail.com>

inge4pres commented:
I rebased and updated with the type-system goodness that we have in main branch.
take a look @hendriknielaender and let me know what you think. 🙏🏼

hendriknielaender commented:
> I rebased and updated with the type-system goodness that we have in main branch.
> take a look @hendriknielaender and let me know what you think. 🙏🏼 


That looks great. 🙏 Thanks for taking care of it.
